playgrounds/react/src/examples/form/components/hooks/use-form-item-layout.tsx

```diff
@@ -0,0 +1,55 @@
+import {
+  Children,
+  ComponentType,
+  isValidElement,
+  JSXElementConstructor,
+  ReactNode,
+  useMemo,
+} from 'react';
+import { CustomRule, CustomRule1 } from '../dummy-validation';
+import { FormItemHint, FormItemLabel } from '../form-item-parts';
+
+type ChildrenArray = ReturnType<typeof Children.toArray>;
+
+const isNodeOfTypes = (types: ComponentType[]) => (node: ReactNode) => isValidElement(node)
+ && types.some((type) => node.type === type);
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function findNodeByTypes(nodes: ChildrenArray, types: JSXElementConstructor<any>[]) {
+  return nodes.find(isNodeOfTypes(types));
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function filterNodesByTypes(nodes: ChildrenArray, types: JSXElementConstructor<any>[]) {
+  return nodes.filter(isNodeOfTypes(types));
+}
+
+export interface FormItemChildrenInfo {
+  editor: ReactNode;
+  label?: ReactNode;
+  hint?: ReactNode;
+  rules: ReactNode[];
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function useFormItemLayout(children: ReactNode, editorTypes: JSXElementConstructor<any>[]) {
+  const {
+    label, hint, editor, rules,
+  } = useMemo<FormItemChildrenInfo>(() => {
+    const childrenArray = Children.toArray(children);
+
+    return {
+      label: findNodeByTypes(childrenArray, [FormItemLabel]),
+      editor: findNodeByTypes(childrenArray, editorTypes),
+      hint: findNodeByTypes(childrenArray, [FormItemHint]),
+      rules: filterNodesByTypes(childrenArray, [CustomRule, CustomRule1]) || [],
```

Comment:
Question.

Maybe one common component for the rules is more suitable here?
Because as far as I understand it clearly, the form item can work only with the rule components types mentioned here.
So, if our client wants to create a custom rule by themself, it must use one of the supported rule types.

Reply:
Ideally, it should be filtered by the base type of the child's props, which should the same for all validation rules. I'll try to implement it in a separate PR